pennylane/workflow/run.py

```diff
@@ -0,0 +1,317 @@
+
+    Raises:
+        pennylane.QuantumFunctionError: If the required package for the specified interface is not installed.
+    """
+    interface = resolved_execution_config.interface
+    grad_on_execution = resolved_execution_config.grad_on_execution
+    device_vjp = resolved_execution_config.use_device_jacobian_product
+    try:
+        if interface == Interface.AUTOGRAD:
+            from .interfaces.autograd import autograd_execute as ml_boundary
+
+        elif interface == Interface.TF_AUTOGRAPH:
+            from .interfaces.tensorflow_autograph import execute as ml_boundary
+
+            ml_boundary = partial(ml_boundary, grad_on_execution=grad_on_execution)
+
+        elif interface == Interface.TF:
+            from .interfaces.tensorflow import tf_execute as full_ml_boundary
+
+            ml_boundary = partial(full_ml_boundary, differentiable=differentiable)
+
+        elif interface == Interface.TORCH:
+            from .interfaces.torch import execute as ml_boundary
+
+        elif interface == Interface.JAX_JIT:
+            from .interfaces.jax_jit import jax_jit_jvp_execute as ml_boundary
+
+        else:  # interface is jax
+            if device_vjp:
+                from .interfaces.jax_jit import jax_jit_vjp_execute as ml_boundary
+            else:
+                from .interfaces.jax import jax_jvp_execute as ml_boundary
+
+    except ImportError as e:  # pragma: no cover
+        raise qml.QuantumFunctionError(
+            f"{interface} not found. Please install the latest "
+            f"version of {interface} to enable the '{interface}' interface."
+        ) from e
+
+    return ml_boundary
+
+
+def _make_inner_execute(device, inner_transform, execution_config=None) -> Callable:
+    """Construct the function responsible for executing quantum tapes within a ML framework boundary
+    for first-order derivatives.
+
+    The process involves performing device-specific preprocessing on the tapes (for new devices) or
+    device expansion (for legacy devices).
+
+    For higher-order derivatives, this function will delegate to another ML framework execution.
+    """
+
+    def inner_execute(tapes: "qml.tape.QuantumScriptBatch", **_) -> ResultBatch:
+        """Execution that occurs within a ML framework boundary.
+
+        Closure Variables:
+            expand_fn (Callable[[QuantumTape], QuantumTape]): A device preprocessing step
+            device (qml.devices.Device)
+        """
+
+        transformed_tapes, transform_post_processing = inner_transform(tapes)
+
+        if transformed_tapes:
+            results = device.execute(transformed_tapes, execution_config=execution_config)
+        else:
+            results = ()
+
+        return transform_post_processing(results)
+
+    return inner_execute
+
+
+# pylint: disable=too-many-branches
+def run(
+    tapes: "qml.tape.QuantumScriptBatch",
+    device: "qml.devices.Device",
+    resolved_execution_config: "qml.devices.ExecutionConfig",
```

Comment:
I know I was the one to say `resolved_execution_config` before, but looking at this now I'm leaning toward just having it be `execution_config` for brevity's sake.  Then in the docstring we just highlight that it must be fully determined.

Reply:
Any objects to just using `config` instead?